### PR TITLE
Add ability to map labels to range 0 to N

### PR DESCRIPTION
**Pull request by nickwp**
Adds the ability to map arbitrary class labels from and H5Dataset into the range from 0 to N, where N is the number of labels in the set. This allows training a classifier for any set of labels, to work around the PyTorch requirement that labels run from 0 to N.

To use this option, the classifer engine should set the config parameter `label_set` in the usual Hydra way for a list. If set to None (the default if `label_set` is not specified) then the class labels must already be the range 0 to N. Otherwise, the set of labels can be specified by, e.g.:

```
label_set:
 - 1
 - 2
```

The softmax output columns will correspond to the set of labels provided, in the order they're provided (rows correspond to events). If no label set is provided, and labels are 0 to N, then the columns correspond to these values (i.e. no change from before).

### CODE REVIEW

**Comment by nickwp** [comment]
This could be extended to resolve the issue where the number of classes needs to be set in several config options (for the different parts of the model) but since that needs more significant code changes, probably best to keep this simple and resolve that in another PR.

**Comment by nickwp** [comment]
Successfully tested this with a run using electrons and muons (i.e. labels 1 and 2), which previously was not possible unless changing the labels defined in the data itself.

**Review by pdeperio — COMMENTED** [review]
is there any way to automate this setting from the input data format?

**Comment by nickwp** [comment]
> is there any way to automate this setting from the input data format?

Yeah, I thought about ways to do this. It's possible, but a couple of reasons I decided not to, at least for now:

- Currently we have a single h5 file with all classes and for 2-class training we only use a subset to actually train on. The way PyTorch separates the dataset class (which processes the data, given which events to load) from the dataloader (which handles which events to load from the dataset) means that the dataset itself doesn't know which subset of classes are being used.
- If the set of classes is not 0 to N (and the labels don't even need to be numeric at all), then we need to decide how to order the columns of the softmax outputs. The simple way is to order them in the order that the class labels are listed in the configuration. If the set of classes is chosen automatically from the dataset, it might not be clear which class corresponds to which column of the softmax array.

I think both of these are solvable if we want to, but like with removing the need to specify number of classes in the model config it would be a more major change to the code structure and possibly the output format.

**Review by pdeperio — APPROVED** [review]
